### PR TITLE
3 add dual dirac model tail fitting

### PR DESCRIPTION
**Pull request by capn-freako**
Dear jitter experts,

Please, take a look at the new code here and advise on any missteps, possible improvements, gross misunderstanding of the relevant theory, etc.

Thanks!
-db

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,6 +18,7 @@ dependencies = [
         "PyIBIS-AMI>=4.1.2",
         "pyside2",
         "pyyaml>=6",
+        "qt>=5",
```
Comment (jdpatt) [inline pyproject.toml:21]: Is this a conda package? I'm not sure this exists in pypi and I would have thought pyside would grab qt dependencies.  Do we need to update from pyside2 to pyside6?

File: src/pybert/pybert.py [hunk 2]
```diff
@@ -152,14 +105,14 @@ class PyBERT(HasTraits):
...
-    do_sweep = Bool(False)  #: Run sweeps? (Default = False)
-    debug = Bool(False)  #: Send log messages to terminal, as well as console, when True. (Default = False)
+    nspb = Range(low=2, high=256, value=32)  #: Signal vector samples per bit.
+    mod_type   = List([0])                   #: 0 = NRZ; 1 = Duo-binary; 2 = PAM-4
+    num_sweeps = Int(1)                      #: Number of sweeps to run.
+    sweep_num  = Int(1)
+    sweep_aves = Int(1)
+    do_sweep   = Bool(False)  #: Run sweeps? (Default = False)
+    debug      = Bool(False)  #: Send log messages to terminal, as well as console, when True. (Default = False)
+    rel_thresh = Float(3.0)   #: Spectral threshold for identifying periodic components (sigma). (Default = 3.0)
```
Comment (jdpatt) [inline src/pybert/pybert.py:115]: Renaming the attributes that get saved into the config such as thresh -> rel_thresh breaks the current loading/saving implementation.  Need to make the same change here: https://github.com/capn-freako/PyBERT/blob/master/src/pybert/configuration.py#L151

File: PyBERT.sublime-project [hunk 3]
```diff
@@ -2,7 +2,13 @@
 	"folders":
 	[
 		{
-			"path": "."
-		}
+			"path": ".",
+		},
+		{
+			"path": "/Users/dbanas/Documents/gitit",
```
Comment (jdpatt) [inline PyBERT.sublime-project:8]: There are your local personal paths.  Doesn't bother me having it in (vscode user myself) but if someone tried to open pybert with sublime; it'd fail for them.

File: src/pybert/pybert.py (hunk 2 above)
Comment (jdpatt) [reply src/pybert/pybert.py:115]: I'm okay with the rename just another note that we don't have any way to update our configuration to a new format yet; so will break older configurations.

**Review by jdpatt — CHANGES_REQUESTED** [review]
After making the change in configuration.py most the tests pass when I run `pytest tests/`.  I still have a failure in `test_adapt` that says DFE is unstable.

```

====================================================== FAILURES =======================================================
________________________________________________ TestBasic.test_adapt _________________________________________________

self = <test_basic.TestBasic object at 0x000001D36A151420>, dut = <pybert.pybert.PyBERT object at 0x000001D367C3FFB0>

    def test_adapt(self, dut):
        """Test DFE lock, by ensuring that last 20% of all coefficient vectors
        are stable to within +/-20% of their mean."""
        _weights = dut.adaptation  # rows = step; cols = tap
        _ws = np.array(list(zip(*_weights[4 * len(_weights) // 5 :])))  # zip(*x) = unzip(x)
        _means = list(map(lambda xs: sum(xs) / len(xs), _ws))
>       assert all(
            map(lambda pr: all(abs(pr[0] - pr[1]) / pr[1] < 0.2), zip(_ws, _means))
        ), "DFE adaptation is unstable!"
E       AssertionError: DFE adaptation is unstable!
E       assert False
E        +  where False = all(<map object at 0x000001D36C8F4250>)
E        +    where <map object at 0x000001D36C8F4250> = map(<function TestBasic.test_adapt.<locals>.<lambda> at 0x000001D36C8CDE10>, <zip object at 0x000001D36C8F1B80>)
E        +      where <zip object at 0x000001D36C8F1B80> = zip(array([[0.30765935, 0.30765935, 0.30765935, ..., 0.30279544, 0.30279544,\n        0.30248771],\n       [0.08838273, 0.08838273, 0.08838273, ..., 0.08877971, 0.08877971,\n        0.08866555],\n       [0.04312577, 0.04312577, 0.04312577, ..., 0.04235612, 0.04235612,\n        0.04129169],\n       [0.02514349, 0.02514349, 0.02514349, ..., 0.02461879, 0.02461879,\n        0.02367616],\n       [0.01762905, 0.01762905, 0.01762905, ..., 0.01639309, 0.01639309,\n        0.01316204]]), [0.30527648187497186, 0.08908681662887283, 0.04249113482865134, 0.025051993298613105, 0.016615732360105984])

tests\test_basic.py:54: AssertionError
```

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-  - repo: https://github.com/psf/black
-    rev: 23.11.0
-    hooks:
-      - id: black
-  - repo: https://github.com/pycqa/isort
-    rev: 5.12.0
-    hooks:
-      - id: isort
-        args: ["--profile", "black", "--filter-files"]
+  # - repo: https://github.com/psf/black
```
Comment (jdpatt) [inline .pre-commit-config.yaml:2]: We emailed about commenting this out and acknowledged.  Note that I'll still plan to use black formatting when I create PRs.

File: src/pybert/pybert.py (hunk 2 above)
Comment (capn-freako) [reply src/pybert/pybert.py:115]: Whoops! I'll change that back to `thresh`.
Were there any others?

File: pyproject.toml (hunk 1 above)
Comment (capn-freako) [reply pyproject.toml:21]: So, this one's been problematic for a while now.
I've removed it a few times over the last couple of years, for exactly the reasons you're pointing out.
But, each time I do, I get a flurry of complaints from Windows users about failed installations.
(I've got 3 or 4 of those now.)
And, sure enough, replacing it seems to always fix the problem.

Wrt/ `PySide6`, I tried to update to it a few months ago and it was a disaster.
I plan to revisit this next year, but would like to get the next release out before the end of this year.

File: PyBERT.sublime-project (hunk 3 above)
Comment (capn-freako) [reply PyBERT.sublime-project:8]: Adding this file was a mistake.
I'll remove it.

File: src/pybert/pybert.py (hunk 2 above)
Comment (jdpatt) [reply src/pybert/pybert.py:115]: I think just the one attribute was changed.

**Comment by capn-freako** [comment]
> After making the change in configuration.py most the tests pass when I run `pytest tests/`

Can you explain this? I have a bunch of errors when I run `purest tests/`. And I'm wondering if I need to make whatever changes you made.

**Comment by capn-freako** [comment]
> I still have a failure in `test_adapt` that says DFE is unstable.

I confirm your finding. But, it's sporadic. If I run 10 testing trials, only 1 or 2 fail. The rest succeed.
And the data that get dumped by the assertion when it does fail look like a passing case.
Also, running several runs back to back in the GUI reveal a rock solid DFE adaptation in all cases.
So, I think this is a false failure.
I'm going to go ahead and push my recent changes.

**Review by capn-freako — COMMENTED** [review]
I have confirmed @jdpatt 's finding, re: DFE adaptation failure, but found that it is only sporadic, occurring roughly 10-20% of the time. Also, it seems to be a red herring, as both the data reported in the exception, as well as visual inspection when running the GUI, reveal the DFE adaptation to be consistently rock solid.

**Comment by capn-freako** [comment]
Thanks, @jdpatt !

Happy Holidays,
-db

**Pull request merged**